water/lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -125,6 +127,96 @@ materializeAffine(Location loc, ArrayRef<Attribute> symbols, AffineMap map,
       }
       continue;
     }
+
+    if (auto iterSymbol = dyn_cast<wave::WaveIterSymbolAttr>(attr)) {
+      // Check if we're inside an scf.for loop that corresponds to this
+      // iteration symbol.
+      Block *currentBlock = rewriter.getInsertionBlock();
+      scf::ForOp parentFor = nullptr;
+
+      // Look for a parent scf.for operation.
+      Operation *current = currentBlock->getParentOp();
+      while (current) {
+        if (auto forOp = dyn_cast<scf::ForOp>(current)) {
+          parentFor = forOp;
+          break;
+        }
+        current = current->getParentOp();
+      }
+
+      if (!parentFor) {
+        // Check if we're inside a wave.iterate instead.
+        Operation *currentOp = currentBlock->getParentOp();
+        while (currentOp) {
+          if (isa<wave::IterateOp>(currentOp)) {
+            return rewriter.notifyMatchFailure(
+                loc, "iteration symbol found inside wave.iterate - "
+                     "please run lower-wave-control-flow pass first");
+          }
+          currentOp = currentOp->getParentOp();
+        }
+
+        return rewriter.notifyMatchFailure(
+            loc, "iteration symbol found but no iteration context available");
+      }
+
+      // Get the induction variable and compute iteration value.
+      Value inductionVar = parentFor.getInductionVar();
+
+      // We need the tile size to compute iteration_value = induction_var *
+      // tile_size Get this from the function's tiling constraints.
+      func::FuncOp parentFunc = current->getParentOfType<func::FuncOp>();
+      if (!parentFunc) {
+        return rewriter.notifyMatchFailure(loc, "no parent function found");
+      }
+
+      auto hyperparamAttr =
+          parentFunc->getAttrOfType<wave::WaveHyperparameterAttr>(
+              wave::WaveDialect::kHyperparameterAttrName);
```

Comment:
There is absolutely no guarantee there is a parent `func.func` specifically or that hyperparameters are attached to a `func.func`. Just walk the parents until you see the attribute. Even better if the attribute is passed down this function from the pass somehow.

Reply:
I double-checked, and we already do this in this pass! Hyperparams are available from the type converter and are even used below in this file.